test/SlowTests/Issues/RavenDB_25412.cs

```diff
@@ -0,0 +1,231 @@
+        var networkGate = new AsyncGate();
+
+        var forTesting = sinkDb.ReplicationLoader.ForTestingPurposesOnly();
+
+        // Inject the wrapper to simulate network hangs on the socket stream
+        forTesting.WrapIncomingReplicationStream = innerStream => new SmartHangingStreamWrapper(innerStream, networkGate);
+
+        await hubStore.Maintenance.SendAsync(new PutPullReplicationAsHubOperation(new PullReplicationDefinition(pullReplicationName)));
+        await sinkStore.Maintenance.SendAsync(new PutConnectionStringOperation<RavenConnectionString>(new RavenConnectionString
+        {
+            Name = connectionStringName,
+            Database = hubStore.Database,
+            TopologyDiscoveryUrls = hubStore.Urls
+        }));
+        await sinkStore.Maintenance.SendAsync(new UpdatePullReplicationAsSinkOperation(new PullReplicationAsSink
+        {
+            Name = pullReplicationName,
+            HubName = pullReplicationName,
+            ConnectionStringName = connectionStringName
+        }));
+
+        // 1. Verify Initial Replication works
+        using (var session = hubStore.OpenAsyncSession())
+        {
+            await session.StoreAsync(new User { Id = "Users/1-A", Name = "Lev" });
+            await session.SaveChangesAsync();
+        }
+        Assert.True(WaitForDocument<User>(sinkStore, "Users/1-A", u => u.Name == "Lev"), "Initial replication failed");
+
+        // 2. Simulate Network Hang
+        // Any subsequent read on the existing stream will now hang indefinitely until we open the gate.
+        networkGate.Close();
+
+        // 3. Write second document to Hub
+        using (var session = hubStore.OpenAsyncSession())
+        {
+            await session.StoreAsync(new User { Id = "Users/2-A", Name = "Lev" });
+            await session.SaveChangesAsync();
+        }
+
+        // 1. The existing connection hangs in ReadAsync.
+        // 2. The finite timeout triggers in AbstractIncomingReplicationHandler.
+        // 3. The socket is disposed.
+        // 4. The Sink attempts to reconnect.
+
+        // Wait enough time for the timeout to trigger and the old connection to die.
+        var timeout = (int)sinkDb.Configuration.Replication.ActiveConnectionTimeout.AsTimeSpan.Add(TimeSpan.FromSeconds(10)).TotalMilliseconds;
+        var isSecondDocReplicated = WaitForDocument<User>(sinkStore, "Users/2-A", u => u.Name == "Lev", timeout);
+        Assert.False(isSecondDocReplicated, "Network gate is not actually closed as intended. Test is broken");
+
+        // Open the gate.
+        // If the Sink has already created a NEW connection, it might be waiting at the gate.
+        // Opening it allows the handshake to proceed.
+        networkGate.Open();
+
+        isSecondDocReplicated = WaitForDocument<User>(sinkStore, "Users/2-A", u => u.Name == "Lev", timeout: timeout);
+        Assert.True(isSecondDocReplicated, "Should have timed out the hung connection, reconnected, and replicated.");
```

Comment:
Can this also pass even without your fix?
you open the gate regardless to what connection you had..